vllm/v1/sample/logits_processor.py

```diff
@@ -501,8 +504,198 @@ def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        elif state["in_think"]:
+            state["think_count"] += 1
+
+        # Transition into end mode if thinking token limit exceeded
+        if state["in_end"]:
+            state["end_count"] += 1
+            if state["end_count"] >= len(self.think_end_token_ids):
+                state.update({"in_end": False, "end_count": 0})
+        else:
+            if state["in_think"] and state["think_count"] \
+                    >= state["thinking_token_budget"]:
+                state.update({
+                    "in_think": False,
+                    "in_end": True,
+                    "end_count": 0
+                })
+
+    def is_argmax_invariant(self) -> bool:
+        """This logits processor can change the outcome of
+        greedy sampling by forcing that the thinking section
+        ends after a certain number of tokens."""
+        return False
+
+    def update_state(self, batch_update: Optional[BatchUpdate]):
+        if batch_update:
+            for (index, params, prompt_tok_ids, output_tok_ids) \
+                in batch_update.added:
+                reasoning_effort = (params.reasoning_effort if isinstance(
+                    params, SamplingParams) else None)
+                thinking_token_budget = (params.thinking_token_budget
+                                         if isinstance(params, SamplingParams)
+                                         else None)
+                resolved_thinking_token_budget = \
+                    self._resolve_thinking_token_budget(
+                        reasoning_effort, thinking_token_budget)
+                if resolved_thinking_token_budget is not None:
+                    self._state[index] = self._init_state_entry(
+                        prompt_tok_ids, resolved_thinking_token_budget)
+                    self._state[index]["output_tok_ids"] = output_tok_ids
+
+            for index in batch_update.removed:
+                self._state.pop(index, {})
+
+            for i1, i2, direction in batch_update.moved:
+                if direction == MoveDirectionality.SWAP:
+                    self._state[i1], self._state[i2] = \
+                        self._state[i2], self._state[i1]
+                else:
+                    self._state[i2] = self._state.pop(i1, {})
+
+        for state in self._state.values():
+            self._update_think_state(state)
+
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        if not self._state:
+            return logits
+
+        batch_size = logits.size(0)
+        self.mask[:batch_size] = False
+
+        for i in range(batch_size):
+            state = self._state.get(i)
+            if state and state["in_end"]:
+                self.mask[i] = True
+                self.force_token_ids[i] = \
+                        self.think_end_token_ids[state["end_count"]]
+
+        current_mask = self.mask[:batch_size]
+        if current_mask.any():
+            logits[current_mask] = -float("inf")
```

Comment:
Line 689 can be avoided, the time taken for this conversion to -float("inf') would be O(batch_szie * vocab_size), which can we significant if a batch size of 128 and above with a vocab of 200k or greater is converted, this could be roughly, 1.2-2.5 ms for bs=512 with vocab=250k +. 
```suggestion
            logits[current_mask,
                   self.force_token_ids[:batch_size][current_mask]] = 1e9 (a really large value for the end thinking token id index)
```